### PR TITLE
Core: Add REST catalog adapter and tests

### PR DESCRIPTION
**Pull request by rdblue**
This adds an implementation of `RESTClient` that translates REST protocol requests to the `Catalog` API. This makes it possible to test the `RESTCatalog` implementation like any other catalog. These tests do not exercise serialization; request and response objects are passed between methods directly.

* `RESTCatalogAdaptor` is a simple translation from `RESTClient` routes to catalog handlers
* `CatalogHandlers` accepts request objects, makes the appropriate calls to a catalog, and produces response objects
* `RESTCatalog` is a `Catalog` implementation that translates to calls to `RESTClient`
* `TestRESTCatalog` runs `CatalogTests` against the `RESTCatalog`

This also implements load table and update table routes, with new tests in `CatalogTests` that are passing for REST and JDBC.

To get tests working, this PR includes a few other changes and fixes:

* Adds `equals` to `CharSequenceSet` for assertions
* Updates `TableMetadataParser` to handle `refs`
* Supports `-1` to signal "last added" in schema, partition spec, and sort order changes
* Fixes `TableMetadata` history timestamps by detecting whether a snapshot was added
* Adds `applyTo(TableMetadata.Builder)` to `MetadataUpdate` to apply changes from the REST protocol on the server side
* Adds requirements classes and validations that are used server side to check the loaded metadata

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 1]
```diff
@@ -319,6 +381,556 @@ public void testNamespaceWithDot() {
     Assert.assertFalse("Namespace should not exist", catalog.namespaceExists(withDot));
   }
 
+  @Test
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:384]: nit: would it make sense to pull these out into a separate PR so that we get those in sooner? Unless this PR won't take too long to be merged of course

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+      for (int pos = 0; pos < parts.size(); pos += 1) {
+        String part = parts.get(pos);
+        if (part.startsWith("{") && part.endsWith("}")) {
+          variablesBuilder.put(pos, part.substring(1, part.length() - 1));
+        } else {
+          requirementsBuilder.put(pos, part);
+        }
+      }
+
+      this.requriedLength = parts.size();
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:121]: nit: typo in `requriedLength`

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+import org.apache.iceberg.rest.requests.CreateTableRequest;
+import org.apache.iceberg.rest.requests.UpdateNamespacePropertiesRequest;
+import org.apache.iceberg.rest.requests.UpdateTableRequest;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.util.Pair;
+
+/**
+ * Adaptor class to translate REST requests into {@link Catalog} API calls.
+ */
+public class RESTCatalogAdapter implements RESTClient {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:49]: would it make sense to have a separate test class for this adapter to make sure routing and other stuff is properly tested?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:49]: I don't think so. I'm actually thinking that this class will probably be moved to tests. What we want to make available is `CatalogHandlers`. The adapter is just a simple way to go between the `RESTClient` API and `CatalogHandlers`.

File: api/src/main/java/org/apache/iceberg/util/CharSequenceSet.java [hunk 4]
```diff
@@ -152,4 +156,28 @@ public boolean removeAll(Collection<?> objects) {
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(wrapperSet);
+  }
+
+  @Override
+  public String toString() {
+    return "CharSequenceSet({" + Streams.stream(iterator()).collect(Collectors.joining(", ")) + "})";
```
[the remark this answers is not in the document]
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/util/CharSequenceSet.java:181]: Good to know!
